src/Databases/DatabaseOnDisk.cpp

```diff
@@ -366,13 +361,24 @@ void DatabaseOnDisk::renameTable(
             throw Exception("Moving tables between databases of different engines is not supported", ErrorCodes::NOT_IMPLEMENTED);
     }
 
+    bool allow_rename_dictionary = from_atomic_to_ordinary || from_ordinary_to_atomic;
+    if (exchange)
+        throw Exception("Tables can be exchanged only in Atomic databases", ErrorCodes::NOT_IMPLEMENTED);
+    if (dictionary && !allow_rename_dictionary)
+        throw Exception("Dictionaries can be renamed only in Atomic databases", ErrorCodes::NOT_IMPLEMENTED);
+
+
     auto table_data_relative_path = getTableDataPath(table_name);
     TableExclusiveLockHolder table_lock;
     String table_metadata_path;
     ASTPtr attach_query;
     /// DatabaseLazy::detachTable may return nullptr even if table exists, so we need tryGetTable for this case.
     StoragePtr table = tryGetTable(table_name, local_context);
+    if (table && table->isDictionary() && !allow_rename_dictionary)
```

Comment:
Quite confusing that we have `bool dictionary` and check `table->isDictionary()`. `table->isDictionary()` means table with `Dictionary` engine?

Reply:
Yes, `dictionary` means that it's `RENAME DICTIONARY` query (not `RENAME TABLE`) and `table->isDictionary()` means that it's table with `Dictionary` engine (it can be either DDL dictionary or just table table that reads from another dictionary). The issue is that `RENAME DICTIONARY` is actually not supported, but after #23436 it become possible to execute `RENAME TABLE` for DDL dictionary (and it works, because we check `dictionary flag` only), but it works incorrectly.